### DISCUSSION TITLE
streams-i2s-webserver_mp3  compile help

### DISCUSSION THREAD

**Discussion by vaporskate** [pschatzmann/arduino-audio-tools#279] [Q&A] [answered]
Using the "streams-i2s-webserver_mp3" example file from the "examples->sandbox" folder I receive the following errors when attempting to compile the file for my ESP-WROOM-32 board in arduino IDE.

`....:  streams-i2s-webserver_mp3.ino.elf section .dram0.bss will not fit in region dram0_0_seg`
`....:  DRAM segment data does not fit.`
`....:  region dram0_0_seg overflowed by 4120 bytes`

Through trial and error I've found that if I turn OFF the memory hack..a.k.a changing `#define USE_MEMORY_HACK 1` in the "config.h" file within liblame library folder to `#define USE_MEMORY_HACK 0`, the errors no longer occur and the file compiles.  However, after compiling and uploading to my ESP32 board I receive an error in the serial monitor that I'm too dumb to solve (see last line of code below).  I understand that there's an error occuring in the "lame.c" file that's getting thrown at line 2740....but how to fix it or how to troubleshoot from here I'm at a loss and was wondering if anyone could help out. Thanks

`[I] I2SConfig.h : 96 - port_no: 0`
`[I] I2SConfig.h : 97 - is_master: Master`
`[I] I2SConfig.h : 98 - sample rate: 22050`
`[I] I2SConfig.h : 99 - bits per sample: 16`
`[I] I2SConfig.h : 100 - number of channels: 2`
`[I] I2SConfig.h : 101 - i2s_format: I2S_MSB_FORMAT`
`[I] I2SConfig.h : 103 - auto_clear:0`
`[I] I2SConfig.h : 110 - buffer_count:10`
`[I] I2SConfig.h : 111 - buffer_size:512`
`[I] I2SConfig.h : 115 - pin_bck: 14`
`[I] I2SConfig.h : 116 - pin_ws: 15`
`[I] I2SConfig.h : 117 - pin_data: 32`
`I2S started`
`[Error] lame.c : 2740 - calloc(1,134792) -> 0 [available: 136152]`

additional note - I've successfully used "streams-i2s-webserver_wav" and it worked like a dream; unfortunately the project i'm working on requires an mp3 stream.

respectfully
vapor

**Comment by pschatzmann** [comment, marked as answer]
I have implemented some additional memory optimizations in the audiotools and LibLame projects.
Please note that a ESP32 needs PSRAM if you want to do WIFI and MP3 encoding at the same time. There is just not enough basic RAM to do both...

**Reply by vaporskate** [reply]
Nailed it.  That was exactly what i needed to know.  I _was_ using an ESP32-WROOM which apparently doesn't have any PSRAM.  I switched to a ESP32-WROVER-E, and it worked perfectly.  Thanks for your help with my question and for building such awesome libraries.
